docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,223 @@
+# Deployment Hooks
+
+## Abstract
+
+A proposal for a deployment code execution API integrated with the deployment lifecycle.
+
+
+## Motivation
+
+Deployment hooks are needed to provide users with a way to execute arbitrary commands necessary to complete a deployment, targeting executables which may be maintained outside the docker image being deployed.
+
+Goals of this design:
+
+1. Identify deployment hook use cases
+2. Define the integration of deployment hooks and the deployment lifecycle
+3. Describe a deployment hook API
+
+
+## Comparison of potential approaches
+
+There are two fundamental approaches to solving each deployment hook use case: existing upstream support for *container lifecycle hooks*, and the externalized *deployment hooks* outlined in this proposal. The following describes the two approaches, and each use case is evaluated in terms of these approaches.
+
+##### Upstream container lifecycle hooks
+
+Kubernetes provides *container lifecycle hooks* for containers within a Pod. Currently, post-start and pre-stop hooks are supported. For deployments, post-start is the most relevant. Because these hooks are implemented by the Kubelet, the post-start hook provides some unique guarantees:
+
+1. The hook is executed synchronously during the pod lifecycle.
+2. The status of the pod is linked to the status and outcome of the hook execution.
+  1. The pod will not enter a ready status until the hook has completed successfully.
+    1. Service endpoints will not be created for the pod until the pod has a ready status.
+  1. If the hook fails, the pod's creation is considered a failure, and the retry behavior is restart-policy driven in the usual way.
+
+Because deployments are represented as replication controllers, lifecycle hooks defined for containers are executed for every container in the replica set for the deployment. This behavior has complexity implications when applied to deployment use cases:
+
+1. The hooks for all pods in the deployment will race, placing a burden on hook authors (e.g., the hooks would generally need to be tolerant of concurrent hook execution and implement manual coordination.)
+
+
+##### Deployment hooks
+
+Deployment hooks implemented as run-once pods external to the Kubelet provide a different set of guarantees:
+
+1. Hooks can be bound to the logical deployment lifecycle, enabling hook executions decoupled from replication mechanics.
+  1. Races can be avoided by defining a hook that executes exactly once per deployment regardless of the replica count.
```

Comment:
It's also cloudier.


> On Mar 12, 2015, at 3:21 PM, Ben Parees <notifications@github.com> wrote:
> 
> In docs/proposals/post-deployment-hooks.md:
> 
> > +2. The status of the pod is linked to the status and outcome of the hook execution.
> > +  1. The pod will not enter a ready status until the hook has completed successfully.
> > +    1. Service endpoints will not be created for the pod until the pod has a ready status.
> > +  1. If the hook fails, the pod's creation is considered a failure, and the retry behavior is restart-policy driven in the usual way.
> > +
> > +Because deployments are represented as replication controllers, lifecycle hooks defined for containers are executed for every container in the replica set for the deployment. This behavior has complexity implications when applied to deployment use cases:
> > +
> > +1. The hooks for all pods in the deployment will race, placing a burden on hook authors (e.g., the hooks would generally need to be tolerant of concurrent hook execution and implement manual coordination.)
> > +
> > +
> > +##### Deployment hooks
> > +
> > +Deployment hooks implemented as run-once pods external to the Kubelet provide a different set of guarantees:
> > +
> > +1. Hooks can be bound to the logical deployment lifecycle, enabling hook executions decoupled from replication mechanics.
> > +  1. Races can be avoided by defining a hook that executes exactly once per deployment regardless of the replica count.
> i'd want at least once. or at least an option for at least once. if you bring up my app w/o having successfully run my db migration i'm not going to be happy. if you run my db migration twice, that might be ok (if i wrote my db migration intelligently)
> 
> —
> Reply to this email directly or view it on GitHub.
>